c2rust-analyze/src/util.rs

```diff
@@ -353,3 +355,44 @@ pub fn is_null_const(constant: Constant) -> bool {
 
 pub trait PhantomLifetime<'a> {}
 impl<'a, T: ?Sized> PhantomLifetime<'a> for T {}
+
+#[derive(Clone, Copy, PartialEq, Eq, Debug, Hash)]
+pub enum TestAttr {
+    /// `#[c2rust_analyze_test::fixed_signature]`: Mark all pointers in the function signature as
+    /// `FIXED`.
+    FixedSignature,
+    /// `#[c2rust_analyze_test::fail_analysis]`: Force an analysis failure for the function.
+    FailAnalysis,
+    /// `#[c2rust_analyze_test::skip_rewrite]`: Skip generating rewrites for the function.
+    SkipRewrite,
+}
+
+impl TestAttr {
+    pub fn name(self) -> &'static str {
+        match self {
+            TestAttr::FixedSignature => "fixed_signature",
+            TestAttr::FailAnalysis => "fail_analysis",
+            TestAttr::SkipRewrite => "skip_rewrite",
+        }
+    }
+}
+
+pub fn has_test_attr(tcx: TyCtxt, ldid: LocalDefId, attr: TestAttr) -> bool {
+    let tool_sym = Symbol::intern("c2rust_analyze_test");
+    let name_sym = Symbol::intern(attr.name());
+
+    for attr in tcx.get_attrs_unchecked(ldid.to_def_id()) {
+        let path = match attr.kind {
+            AttrKind::Normal(ref item, _) => &item.path,
+            AttrKind::DocComment(..) => continue,
+        };
+        let (a, b) = match &path.segments[..] {
+            &[ref a, ref b] => (a, b),
```

Comment:
nit:

```suggestion
            [a, b] => (a, b),
```

Reply:
I mostly avoid using the "match ergonomics" feature - it makes it less clear whether the `match` is binding by reference or by value, which is generally useful to know IMO.

Reply:
Isn't that clear from the `&` in the `&path.segments[..]`?  The `&` and `ref` are basically inverses, right, so doing `&[ref` seems quite redundant.

Reply:
IMO `&[ref a, ref b]` is much more confusing than `[a, b]`.  Match ergonomics are used all over the place, in most idiomatic Rust code and ours.